README.md
# CUEgenerator

A small utility which facilitates creating cue files from track lists.

# How to report bugs

Please create [a new issue](https://github.com/dVaffection/cuegenerator-react/issues) in this repository and leave as many details as possible.

# Documentation

## Track list highlights

- In case if a local performer is absent the global one is used, e.g.
  ![](https://user-images.githubusercontent.com/457097/111331254-b81dff00-863e-11eb-9b80-cc7504f4d6fd.png)

- Performer and title track separators:
  - `' - '` — 45 (hyphen-minus)
  - `' – '` — 8211 (en dash)
  - `' ‒ '` — 8210 (figure dash)
  - `' — '` — 8212 (em dash)
  - `' ― '` — 8213 (horizontal bar)
- Timings recognition:
  - `[08:45] 03. 8 Ball` → `08:45`
  - `01.[18:02] Giuseppe` → `18:02`
  - `10:57 02. Space Manoeuvres` → `10:57`
  - `56:53 T.O.M'` → `56:53`
  - `1:02:28 Mossy` → `62:28`

## Regions list recognition

- Sony Sound Forge format `dd:dd:dd[.,]dd`
- Adobe Audition format `dd:dd:dd:dd`
- Audacity format `ddddd.dddddd`

# For developers

This is a React frontend to the [BFF](https://github.com/DmitryVarennikov/cuegenerator-server).

## How to deploy

GitHub Actions provide 2 workflows which deploy code to the Firebase Hosting

- on a pull request the branch in question is deployed to a temporary channel for a preview
- on push to the `main` branch if commit contains "(release):" string the code is deployed to the live channel

## How to release

A new release happens when a new commit prefixed with *fix* or *feat* is merged to the `main` branch. The logic is described in the `firebase-hosting-merge.yml` file.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### 1.3.1 (2022-07-15)


### Bug Fixes

* automatic releases v6 ([e2b4d09](https://github.com/DmitryVarennikov/cuegenerator-react/commit/e2b4d09792ec669a53e9d81a7464c6a822fd5082))

## [1.3.0](https://github.com/dVaffection/cuegenerator-react/compare/v1.2.1...v1.3.0) (2021-05-17)


### Features

* Keep the previous cue form state ([0b0b64c](https://github.com/dVaffection/cuegenerator-react/commit/0b0b64ccc72daa62de47f6852e1834fd85fdd5cf))


### Bug Fixes

* Change app name to CUEgenerator ([c359579](https://github.com/dVaffection/cuegenerator-react/commit/c359579cfe01aa30c397e31cbbce62501b1b706d)), closes [#27](https://github.com/dVaffection/cuegenerator-react/issues/27)

### [1.2.1](https://github.com/dVaffection/cuegenerator-react/compare/v1.2.0...v1.2.1) (2021-05-01)


### Bug Fixes

* Tweak parser to cast minutes to "XX", e.g. "05:12:03" instead of "5:12:03" ([21f3604](https://github.com/dVaffection/cuegenerator-react/commit/21f360474ccf8f9603eb388198d77e22d9bf7ae9))

## [1.2.0](https://github.com/dVaffection/cuegenerator-react/compare/v1.1.0...v1.2.0) (2021-04-30)


### Features

* Replace double quotes with single quotes in performer and title ([d2451c7](https://github.com/dVaffection/cuegenerator-react/commit/d2451c7c54a08f798c014ce507e30b408dc2e996)), closes [#23](https://github.com/dVaffection/cuegenerator-react/issues/23)


### Bug Fixes

* Uncaught (in promise) TypeError: Failed to fetch in Counter.tsx:12 ([c5fc88e](https://github.com/dVaffection/cuegenerator-react/commit/c5fc88ec274c81889503d39313e4b29757a9d5aa)), closes [#24](https://github.com/dVaffection/cuegenerator-react/issues/24)

## [1.1.0](https://github.com/dVaffection/cuegenerator-react/compare/v1.0.4...v1.1.0) (2021-03-28)


### Features

* Parse time in track list without leading zero for minutes ([8eb37de](https://github.com/dVaffection/cuegenerator-react/commit/8eb37de4d2d9b627750d04bfa9c14bf3973e4055))

### [1.0.4](https://github.com/dVaffection/cuegenerator-react/compare/v1.0.3...v1.0.4) (2021-03-28)


### Bug Fixes

* a downloadable cue file must have the ".cue" extension ([2f928b8](https://github.com/dVaffection/cuegenerator-react/commit/2f928b8d22f4a69ae9a559368e076fcac39a92bd))

### [1.0.3](https://github.com/dVaffection/cuegenerator-react/compare/v1.0.2...v1.0.3) (2021-03-24)

### [1.0.2](https://github.com/dVaffection/cuegenerator-react/compare/v1.0.1...v1.0.2) (2021-03-23)

### 1.0.1 (2021-03-23)


src/Cue/Parser.ts
import { RegionsList, TrackList } from '../types';

export class ParserHelper {
  // that's how we tell performer and title apart
  private titlePerformerSeparators = [
    ' - ', // 45 hyphen-minus
    ' – ', // 8211 en dash
    ' ‒ ', // 8210 figure dash
    ' — ', // 8212 em dash
    ' ― ', // 8213 horizontal bar
  ];
  splitTitlePerformer(value: string) {
    // `foreach` and `switch` are toooooooooo slow!
    let split = [],
      performer = '',
      title = '';

    if (-1 !== value.search(this.titlePerformerSeparators[0])) {
      split = value.split(this.titlePerformerSeparators[0]);
    } else if (-1 !== value.search(this.titlePerformerSeparators[1])) {
      split = value.split(this.titlePerformerSeparators[1]);
    } else if (-1 !== value.search(this.titlePerformerSeparators[2])) {
      split = value.split(this.titlePerformerSeparators[2]);
    } else if (-1 !== value.search(this.titlePerformerSeparators[3])) {
      split = value.split(this.titlePerformerSeparators[3]);
    } else if (-1 !== value.search(this.titlePerformerSeparators[4])) {
      split = value.split(this.titlePerformerSeparators[4]);
    } else {
      split = [value];
    }

    // if string wasn't split yet then we get just a title (performer assumed to be the global one)
    if (1 === split.length) {
      performer = '';
      title = split.shift() || '';
    } else {
      performer = split.shift() || '';
      title = split.join(' ');
    }

    return {
      performer: performer.trim(),
      title: title.trim(),
    };
  }

  separateTime(value: string) {
    var time = '',
      residue = '';

    // ask to increase minutes up to 9999 referred to the "h:m" not "h:m:f" format
    // but I still increased "h:m:f" up to 999 hours just in case
    // https://github.com/dVaffection/cuegenerator/issues/14

    //                          01.     9999:53 | 999:02:28
    var pattern = /^(?:\d{2}\.)?\[?((?:\d{1,3}:)?\d{1,4}:\d{2})\]?.*$/i;
    var matches = value.match(pattern);
    if (matches && matches[1]) {
      time = matches[1].trim();
      residue = value.substring(value.indexOf(matches[1]) + matches[1].length).trim();
    } else {
      residue = value.trim();
    }

    return {
      time: time,
      residue: residue,
    };
  }

  /**
   * Accept time in format either hr:mn:sc or mn:sc
   *
   * @param {String}
   * @returns mn:sc:fr
   */
  castTime(value: string) {
    let castTime = '00:00:00';
    value = value.trim();

    const pattern = /^\d{1,4}:\d{2}:\d{2}$/;
    const matches = value.match(pattern);
    if (matches) {
      const times = value.split(':');
      const hrParsed = parseInt(times[0], 10);
      const mnParsed = parseInt(times[1], 10);
      const sc = times[2].padStart(2, '0');
      const mn = String(hrParsed * 60 + mnParsed).padStart(2, '0');
      castTime = mn + ':' + sc + ':00';
    } else {
      const pattern = /(^\d{1,4}):(\d{2})$/;
      const matches = value.match(pattern);
      if (matches) {
        const mn = matches[1].padStart(2, '0');
        const sc = matches[2].padStart(2, '0');
        castTime = `${mn}:${sc}:00`;
      }
    }

    return castTime;
  }

  cleanOffTime(value: string) {
    var pattern = /^(?:\]? )?(?:\d{2}\)?\.? )?(.*)$/i;
    var matches = value.match(pattern);

    if (matches && matches[1]) {
      value = matches[1];
    }

    return value;
  }

  removeDoubleQuotes(value: string) {
    return value.replace(/"/g, '');
  }

  replaceDoubleQuotes(value: string) {
    return value.replace(/"/g, "'");
  }
}

export default class Parser {
  private helper: ParserHelper;

  constructor(helper: ParserHelper) {
    this.helper = helper;
  }

  parsePerformer(v: string): string {
    return v.trim();
  }
  parseTitle(v: string): string {
    return v.trim();
  }
  parseFileName(v: string): string {
    return v.trim();
  }
  parseTrackList(value: string): TrackList {
    const trackList = [];
    let time, performer, title;

    const contentInLines = value.split('\n');
    for (let i = 0, track = 1; i < contentInLines.length; i++, track++) {
      const row = contentInLines[i].trim();
      if (!row.length) {
        track--;
        continue;
      }

      const performerTitle = this.helper.splitTitlePerformer(row);

      if (performerTitle.performer) {
        const timePerformer = this.helper.separateTime(performerTitle.performer);
        time = this.helper.castTime(timePerformer.time);
        performer = this.helper.cleanOffTime(timePerformer.residue);
        title = performerTitle.title;
      } else {
        performer = '';
        const timeTitle = this.helper.separateTime(performerTitle.title);
        time = this.helper.castTime(timeTitle.time);
        title = this.helper.cleanOffTime(timeTitle.residue);
      }

      performer = this.helper.replaceDoubleQuotes(performer);
      title = this.helper.replaceDoubleQuotes(title);

      trackList.push({
        track,
        performer,
        title,
        time,
      });
    }

    return trackList;
  }

  parseRegionsList(value: string): RegionsList {
    const regionsList: RegionsList = [];
    let matches;
    var contents = value.split('\n');

    for (let i = 0; i < contents.length; i++) {
      const row = contents[i].trim();

      // Audition marker export
      matches = row.match(/(\d+):(\d+)\.(\d+)\t\d+:\d+\.\d+\tdecimal\tCue/i);
      if (null != matches) {
        let mn = Number(matches[1]) || 0;
        const sc = Number(matches[2]) || 0;
        const milliseconds = Number(matches[3]) || 0;

        // frames can not be more than 74, so floor them instead of round
        const fr = Math.floor(parseFloat(0 + '.' + milliseconds) * 75);

        const timeEntry = (mn < 10 ? '0' + mn : mn) + ':' + (sc < 10 ? '0' + sc : sc) + ':' + (fr < 10 ? '0' + fr : fr);
        regionsList.push(timeEntry);

        continue;
      }

      // Soundforge or Audition
      matches = row.match(/(\d{2}:\d{2}:\d{2}[\.,:]\d{2})/i);
      if (null != matches) {
        let time = matches[0].split(':');
        const hr = time.shift() || '0';
        let mn: string | number = time.shift() || '0';

        // frames can be separated by .(dot) or :(colon) or ,(comma)
        if (time.length > 1) {
          var sc = time.shift();
          var fr = time.shift();
        } else {
          let sc_fr = time.shift() || '';
          let sc_fr_split;
          switch (true) {
            case -1 != sc_fr.indexOf('.'):
              sc_fr_split = sc_fr.split('.');
              var sc = sc_fr_split.shift();
              var fr = sc_fr_split.shift();
              break;
            case -1 != sc_fr.indexOf(','):
              sc_fr_split = sc_fr.split(',');
              var sc = sc_fr_split.shift();
              var fr = sc_fr_split.shift();
              break;
          }
        }

        mn = parseInt(mn, 10) + parseInt(hr, 10) * 60;
        const timeEntry = (mn < 10 ? '0' + mn : mn) + ':' + sc + ':' + fr;
        regionsList.push(timeEntry);

        continue;
      }

      // find Nero/Winamp formats mm(m):ss(:|.)ii
      matches = row.match(/(\d{2,3}:\d{2}[\.:]\d{2})/i);
      if (null != matches) {
        var time = matches[0].split(':');
        var mn = time.shift();
        if (time.length == 1) {
          var sc_fr = time[0].split('.');
          var sc = sc_fr.shift();
          var fr = sc_fr.shift();
        } else {
          var sc = time.shift();
          var fr = time.shift();
        }

        const timeEntry = mn + ':' + sc + ':' + fr;
        regionsList.push(timeEntry);

        continue;
      }

      // Audacity
      matches = row.match(/(\d{1,5}).(\d{6})/i);
      if (null != matches) {
        const milliseconds = matches[2];
        const seconds = Number(matches[1]) || 0;
        const minutes = Math.floor(seconds / 60);

        let mn = minutes > 0 ? minutes : 0;
        let sc = seconds % 60;
        // frames can not be more than 74, so floor them instead of round
        const fr = Math.floor(parseFloat(0 + '.' + milliseconds) * 75);

        const timeEntry = (mn < 10 ? '0' + mn : mn) + ':' + (sc < 10 ? '0' + sc : sc) + ':' + (fr < 10 ? '0' + fr : fr);
        regionsList.push(timeEntry);

        continue;
      }

      // try to recognise raw cue timings
      matches = row.match(/(\d{2}:\d{2}:\d{2})/i);
      if (null != matches) {
        time = matches[0].split(':');
        const mn = parseInt(time[0], 10);
        const sc = parseInt(time[1], 10);
        const fr = parseInt(time[2], 10);

        const timeEntry = (mn < 10 ? '0' + mn : mn) + ':' + (sc < 10 ? '0' + sc : sc) + ':' + (fr < 10 ? '0' + fr : fr);
        regionsList.push(timeEntry);

        continue;
      }
    }

    return regionsList;
  }
}
